File: BrownianMotion/demo2022_repl.py
from random import random, randint
from math import fabs, sqrt

dim = 3 # cuantas dimensiones
dur = 5000
replicas = 30 # cuantas veces
mayores = []
regresos = []
print(replicas, 'caminatas de', dur, 'pasos en', dim, 'dimensiones')
for replica in range(replicas): 
    pos = [ 0 ] * dim
    mayor = 0
    re = 0
    for paso in range(dur):
        eje = randint(0, dim - 1) # elegir eje (van 0, 1, ..., dim - 1)
        if random() < 0.5:
            pos[eje] += 1
        else:
            pos[eje] -= 1
        if all([ p == 0 for p in pos ]): # si todos son ceros
            re += 1 # regreso al origen
        mayor = max(mayor, sqrt(sum([ p**2 for p in pos ])))
    regresos.append(re)
    mayores.append(mayor)
print('Regresos:')
print('Menor cantidad de regresos: ', min(regresos))
print('Mayor cantidad de regresos: ', max(regresos))
print('Cantidad de regresos en promedio: ', sum(regresos) / len(regresos))
print('Distancias maximales:')
print('Menor valor: ', min(mayores))
print('Mayor valor: ', max(mayores))
print('Promedio: ', sum(mayores) / len(mayores))

File: BrownianMotion/demo2022_repl.R
dim = 5 # cuantas dimensiones
dur = 1000
replicas = 50
cat(replicas, 'caminatas de', dur, 'pasos en', dim, 'dimensiones', '\n')
dist = numeric() # lista sin nada
reg = numeric() # lista sin nada
for (r in 1:replicas) {
    pos = rep(0, 5)
    mayor = 0
    regresos = 0
    for (paso in 1:dur) {    
        eje = round(runif(1, 1, dim)) 
        if (runif(1) < 0.5) { # subir en ese eje
            pos[eje] = pos[eje] + 1
        } else { # bajar en ese eje
            pos[eje] = pos[eje] - 1
        }
        if (all(pos == 0)) { # si todos son ceros
            regresos = regresos + 1 # regreso al origen
        }
        mayor = max(mayor, sum(abs(pos))) # manhattan
    }
    dist = c(dist, mayor) # agregar la nueva mayor distancia
    reg = c(reg, regresos) # agregar el nuevo contador de regresos
}
print('Regresos')
summary(reg)
print('Mayores distancias (Manhattan)')
summary(dist)
